strider/query_planner.py

```diff
@@ -277,28 +230,107 @@ async def find_valid_permutations(
+    for edge_id, edge in qgraph["edges"].items():
+        ograph["edges"][edge_id] = {
+            "source": edge["subject"],
+            "predicate": [f"-{p}->" for p in edge["predicate"]],
+            "target": edge["object"],
+        }
+        if reverse:
+            ograph["edges"][edge_id + REVERSE_EDGE_SUFFIX] = {
+                "source": edge["object"],
+                "predicate": [f"<-{p}-" for p in edge["predicate"]],
+                "target": edge["subject"],
+            }
+
+    return ograph
+
+
+async def filter_categories_predicates(operation_graph, operation_kp_map):
+    """Filter out categories and predicates that cannot be found using KPs."""
+    for node in operation_graph['nodes'].values():
+        node['filtered_categories'] = set()
+    for edge in operation_graph['edges'].values():
+        edge['filtered_predicates'] = set()
+
+    for edge in operation_graph['edges'].values():
+
+        source = operation_graph['nodes'][edge['source']]
+        target = operation_graph['nodes'][edge['target']]
+
+        for op in operation_kp_map:
+            if (
+                    op.source_category in source["category"]
+                    and op.target_category in target["category"]
+                    and op.edge_predicate in edge["predicate"]
+            ):
+                source['filtered_categories'].add(op.source_category)
+                target['filtered_categories'].add(op.target_category)
+                edge['filtered_predicates'].add(op.edge_predicate)
 
     # Filter down node categories using those
     # we have recieved from the KP registry
     # to limit the number of permutations.
 
     # We only filter nodes that are touching at least one edge ("connected")
     connected_nodes = set()
-    for edge in expanded_qg['edges'].values():
-        connected_nodes.add(edge['subject'])
-        connected_nodes.add(edge['object'])
+    for edge in operation_graph['edges'].values():
+        connected_nodes.add(edge['source'])
+        connected_nodes.add(edge['target'])
     for node_id in connected_nodes:
-        node = expanded_qg['nodes'][node_id]
+        node = operation_graph['nodes'][node_id]
         if 'category' not in node:
             continue
         node['category'] = list(node.pop('filtered_categories'))
 
     # Also filter edge predicates as well
-    for edge in expanded_qg['edges'].values():
+    for edge in operation_graph['edges'].values():
         edge['predicate'] = list(edge.pop('filtered_predicates'))
+    return None
+
+
+# pylint: disable=too-many-locals
+async def find_valid_permutations(
+        operation_graph,
+        kp_registry: Registry = None,
+        logger: logging.Logger = LOGGER,
+) -> list[dict]:
+    """
+    Given a query graph, generate a list of query graphs
+    that are solvable, and annotate those with a 'request_kp' property
+    that shows which KPs to contact to get results
```

Comment:
Docstring and function name should be updated. Maybe find_complete_operation_graphs or something?